components/places/src/bookmark_sync/incoming.rs

```diff
@@ -38,31 +40,21 @@ impl<'a> IncomingApplicator<'a> {
         // This is a little odd - we need to go via serde_json::Value so we can
         // work out what we want to deserialize it as.
         let timestamp = record.envelope.modified;
-        let json_content = record.into_content::<serde_json::Value>();
+        let json_content = record.into_content::<BookmarkItemRecord>();
```

Comment:
question: Will this still parse and flag records with invalid (wrong type or missing) properties, or will those now fall through to the `IncomingKind::Malformed` case? And, is that a case we should still be worried about?

If memory serves right, we switched to treating records as dynamically-checked `JsonValue` bags in #1661, which is also how Desktop behaves. We don't fail the entire sync anymore, but I'm wondering if we could end up in a situation where Desktop would fix up (replace / reupload) an invalid record, but Rust Places would treat that record as malformed and skip it.

Maybe keeping an eye on the `malformed-incoming-bookmark` probe is enough...

Reply:
I mentioned this in [the commit](https://github.com/mozilla/application-services/pull/5438/commits/8e1c752d08e182df52585c063b47f042414f1e68), but I forgot to call it out in the PR.  I think you're analysis is correct and it means we will skip over records with incorrect types / missing required fields.  I think that's okay, given the payload evolution discussions, but I wonder what other's think.

BTW, I also forgot to call out Mark's excellent new sync15 code that makes this all work.  It was a real joy doing that refactor.

Reply:
> it means we will skip over records with incorrect types / missing required fields.

I'm a little torn here - bookmarks is somewhat special in that it has a fuzzy concept of requiring a kind of atomicity. Eg, [allow_failed](https://github.com/mozilla/application-services/blob/main/components/sync15/src/client/request.rs#L188) when posting exists only for bookmarks. The main concern is the possibility of creating an invalid tree - eg, if the POST creating a parent node failed, we don't want to allow the children. The same concern applies for incoming records - if the parent is malformed, we really can't skip it. So the decision to just give up was kinda explicit, but mainly in a "let's not think about this now" kind of way.

dogear *should* prevent us actually having a bad tree locally though - but if it really is just this device that considers the record malformed and dogear reparents items because of it, then that contagion ends up spreading to other devices too.

All that said though, one malformed record preventing bookmarks syncing *at all* on this device certainly doesn't seem an ideal outcome either. So there's no correct answer here, it's all a tradeoff of what we think might go wrong.

I'd love to hear what others think here.

Reply:
*sigh* - I really should not try and comment without more context. I just deleted the above comment which totally missed the point - we do already skip malformed records as Lina noted. I'm actually not sure that was the right call for reasons noted by Lina and in my deleted comment. I wonder if we can change the "unpack" functions to take the json as `mut &` and them call [take](https://docs.rs/serde_json/latest/serde_json/value/enum.Value.html#method.take), then any non-null values left are unknown?

(Or we assert that the last 5 years *has* cleaned up things such that we can trust data better, however issues like https://mozilla.sentry.io/issues/3796658061 worry me - sentry hates me, but we see some reports with text `Invalid incoming T: invalid type: string "1676479234.51", expected i64` as well as null confusion in `Invalid incoming T: invalid type: null, expected a string`. We can be reasonably sure that the `i64` problem is *not* bookmarks, because `unpack_optional_i64` handles that case - so those kinds of records would regress here)

Reply:
Actually, I don't think that will work for this purpose since the functions were doing much more complicated things like trying to convert strings to ints and also marking when a record needs to be reuploaded.